[experimental] Cleanup and refactor

l1ttleO opened the pull request.

ElgarL commented on src/org/anjocaido/groupmanager/GroupManager.java:
```diff
@@ -151,14 +148,17 @@ public void onDisable(boolean restarting) {
 				worldsHolder.saveChanges(false);
 			} catch (IllegalStateException ex) {
-				GroupManager.logger.warning(ex.getMessage());
+				GroupManager.logger.log(Level.WARNING, ex.getMessage());
```
This is functionally identical to the code it replaces. The only difference is this requires an import.

ElgarL commented on src/org/anjocaido/groupmanager/GroupManager.java:
```diff
@@ -167,22 +167,10 @@ public void onDisable(boolean restarting) {
-
-		if (!restarting)
-			GroupManager.logger.removeHandler(ch);
+		GroupManager.logger.log(Level.INFO, String.format(Messages.getString("GroupManager.DISABLED"), pdfFile.getVersion())); //$NON-NLS-1$
```
This is functionally identical to the code it replaces. The only difference is this requires an import.

ElgarL commented on src/org/anjocaido/groupmanager/dataholder/worlds/WorldsHolder.java:
```diff
@@ -463,9 +449,7 @@ public boolean isWorldKnown(String worldName) {
 	public boolean hasOwnData(String worldName) {
 
 		String key = worldName.toLowerCase();
-		synchronized(worldsData) {
```
Be very careful removing synchronization.

ElgarL commented on src/org/anjocaido/groupmanager/events/GMGroupEvent.java:
```diff
@@ -91,7 +93,7 @@ public void schedule(final GMGroupEvent event) {
 		synchronized (GroupManager.getGMEventHandler().getServer()) {
 			if (GroupManager.getGMEventHandler().getServer().getScheduler().scheduleSyncDelayedTask(GroupManager.getGMEventHandler().getPlugin(), () -> GroupManager.getGMEventHandler().getServer().getPluginManager().callEvent(event), 1) == -1)
-				GroupManager.logger.warning("Could not schedule GM Event.");
+				GroupManager.logger.log(Level.WARNING, "Could not schedule GM Event.");
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/events/GMSystemEvent.java:
```diff
@@ -68,7 +70,7 @@ public void schedule(final GMSystemEvent event) {
 		synchronized (GroupManager.getGMEventHandler().getServer()) {
 			if (GroupManager.getGMEventHandler().getServer().getScheduler().scheduleSyncDelayedTask(GroupManager.getGMEventHandler().getPlugin(), () -> GroupManager.getGMEventHandler().getServer().getPluginManager().callEvent(event), 1) == -1)
-				GroupManager.logger.warning("Could not schedule GM Event.");
+				GroupManager.logger.log(Level.WARNING, "Could not schedule GM Event.");
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/events/GMUserEvent.java:
```diff
@@ -93,7 +95,7 @@ public void schedule(final GMUserEvent event) {
 		synchronized (GroupManager.getGMEventHandler().getServer()) {
 			if (GroupManager.getGMEventHandler().getServer().getScheduler().scheduleSyncDelayedTask(GroupManager.getGMEventHandler().getPlugin(), () -> GroupManager.getGMEventHandler().getServer().getPluginManager().callEvent(event), 1) == -1)
-				GroupManager.logger.warning("Could not schedule GM Event.");
+				GroupManager.logger.log(Level.WARNING, "Could not schedule GM Event.");
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/events/GMWorldListener.java:
```diff
@@ -72,9 +74,9 @@ public void onWorldInit(WorldInitEvent event) {
 			if (plugin.getWorldsHolder().isWorldKnown(worldName)) {
-				GroupManager.logger.info(Messages.getString("GMWorldListener.CONFIGURE_NEW_WORLD")); //$NON-NLS-1$
+				GroupManager.logger.log(Level.INFO, Messages.getString("GMWorldListener.CONFIGURE_NEW_WORLD")); //$NON-NLS-1$
 			} else
-				GroupManager.logger.severe(Messages.getString("GMWorldListener.ERROR_UNRECOGNISED_WORLD")); //$NON-NLS-1$
+				GroupManager.logger.log(Level.SEVERE, (Messages.getString("GMWorldListener.ERROR_UNRECOGNISED_WORLD"))); //$NON-NLS-1$
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/localization/Messages.java:
```diff
@@ -57,7 +58,7 @@ public static void setLanguage() {
 		} catch (Exception ex) {
 			// Invalid name, use default.
-			GroupManager.logger.warning("Missing or corrupt 'language' node. Using default settings");
+			GroupManager.logger.log(Level.WARNING, "Missing or corrupt 'language' node. Using default settings");
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/permissions/AnjoPermissionsHandler.java:
```diff
@@ -208,8 +210,8 @@ private boolean wildcardNegation(Set<String> playerPermArray, String node) {
-			if (playerPermArray.contains("-" + builder.toString())) {
-				GroupManager.logger.fine("Wildcard Negation found for " + node);
+			if (playerPermArray.contains("-" + builder)) {
+				GroupManager.logger.log(Level.FINE, "Wildcard Negation found for " + node);
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/permissions/AnjoPermissionsHandler.java:
```diff
@@ -220,7 +222,7 @@ private boolean wildcardNegation(Set<String> playerPermArray, String node) {
 		 * No negated parent found so return false.
 		 */
-		GroupManager.logger.fine("No Negation found for " + node);
+		GroupManager.logger.log(Level.FINE, "No Negation found for " + node);
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/permissions/BukkitPermissions.java:
```diff
@@ -234,15 +235,15 @@ public void updatePermissions(Player player, String world) {
 						player.updateCommands();
 					} else {
-						Bukkit.getScheduler().runTask(plugin, () -> { player.updateCommands(); });
+						Bukkit.getScheduler().runTask(plugin, player::updateCommands);
```
Functionally the same.

l1ttleO replied:
some intellij warning i dont remember

ElgarL commented on src/org/anjocaido/groupmanager/permissions/BukkitPermissions.java:
```diff
@@ -234,15 +235,15 @@ public void updatePermissions(Player player, String world) {
 		}
 
-		GroupManager.logger.finest("Attachment updated for: " + player.getName());
+		GroupManager.logger.log(Level.FINEST, "Attachment updated for: " + player.getName());
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/permissions/BukkitPermissions.java:
```diff
@@ -350,10 +351,8 @@ public void updatePermissions(Player player, String world) {
-			return alreadyVisited;
 
-		return null;
+		return !alreadyVisited.isEmpty() ? alreadyVisited : null;
```
you could remove the ! and reverse this logic to make it easier to read

`return alreadyVisited.isEmpty() ? null : alreadyVisited ;`

l1ttleO replied:
will do

ElgarL commented on src/org/anjocaido/groupmanager/permissions/BukkitPermissions.java:
```diff
@@ -493,7 +481,7 @@ private void playerJoin(PlayerEvent event) {
 			Player player = event.getPlayer();
 
-			GroupManager.logger.finest("Player Join event: " + player.getName());
+			GroupManager.logger.log(Level.FINEST, "Player Join event: " + player.getName());
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/permissions/BukkitPermissions.java:
```diff
@@ -384,15 +381,6 @@ public void updatePermissions(Player player, String world) {
 
 		List<String> perms = new ArrayList<>();
 
-		/*
-		 * // All permissions registered with Bukkit for this player
-		 * PermissionAttachment attachment = this.attachments.get(player);
-		 *
-		 * // List perms for this player perms.add("Attachment Permissions:");
-		 * for(Map.Entry<String, Boolean> entry :
-		 * attachment.getPermissions().entrySet()){ perms.add(" " +
-		 * entry.getKey() + " = " + entry.getValue()); }
-		 */
```
Reason for removal?

ElgarL commented on src/org/anjocaido/groupmanager/storage/Yaml.java:
```diff
@@ -218,7 +218,7 @@ public void loadAllSearchedWorlds() {
 		for (File folder : worldsFolder.listFiles()) {
 			if (folder.isDirectory() && !folder.getName().startsWith(".")) { //$NON-NLS-1$
-				GroupManager.logger.info(String.format(Messages.getString("WorldsHolder.WORLD_FOUND"), folder.getName())); //$NON-NLS-1$
+				GroupManager.logger.log(Level.INFO, String.format(Messages.getString("WorldsHolder.WORLD_FOUND"), folder.getName())); //$NON-NLS-1$
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/storage/Yaml.java:
```diff
@@ -404,7 +404,7 @@ public void loadGroups(WorldDataHolder dataHolder) throws IOException {
 								thisGrp.addTimedPermission(split[0], Long.parseLong(split[1]));
 							} catch (Exception e) {
-								GroupManager.logger.warning("TimedPermission error: " + nodeData.toString());
+								GroupManager.logger.log(Level.WARNING, "TimedPermission error: " + nodeData);
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/storage/Yaml.java:
```diff
@@ -433,8 +433,8 @@ public void loadGroups(WorldDataHolder dataHolder) throws IOException {
 				 */
-				GroupManager.logger.warning(String.format(Messages.getString("WorldDatHolder.WARN_GROUP_NO_INFO"), thisGrp.getName()));
-				GroupManager.logger.warning(Messages.getString("WorldDatHolder.WARN_USING_DEFAULT") + groupsFile.getPath());
+				GroupManager.logger.log(Level.WARNING, (String.format(Messages.getString("WorldDatHolder.WARN_GROUP_NO_INFO"), thisGrp.getName())));
+				GroupManager.logger.log(Level.WARNING, Messages.getString("WorldDatHolder.WARN_USING_DEFAULT") + groupsFile.getPath());
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/storage/Yaml.java:
```diff
@@ -613,7 +613,7 @@ public void loadUsers(WorldDataHolder dataHolder) throws IOException {
 											thisUser.addTimedPermission(split[0], Long.parseLong(split[1]));
 										} catch (Exception e) {
-											GroupManager.logger.warning("TimedPermission error: " + o.toString());
+											GroupManager.logger.log(Level.WARNING, "TimedPermission error: " + o);
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/storage/Yaml.java:
```diff
@@ -634,7 +634,7 @@ public void loadUsers(WorldDataHolder dataHolder) throws IOException {
 										thisUser.addTimedPermission(split[0], Long.parseLong(split[1]));
 									} catch (Exception e) {
-										GroupManager.logger.warning("TimedPermission error: " + nodeData.toString());
+										GroupManager.logger.log(Level.WARNING, "TimedPermission error: " + nodeData);
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/storage/Yaml.java:
```diff
@@ -674,7 +674,7 @@ public void loadUsers(WorldDataHolder dataHolder) throws IOException {
 					Group hisGroup = dataHolder.getGroup(nodeData.toString());
 					if (hisGroup == null) {
-						GroupManager.logger.warning(String.format(Messages.getString("WorldDatHolder.WARN_NO_GROUP_STATED"), thisUserNode.get("group").toString(), thisUser.getLastName(), dataHolder.getDefaultGroup().getName(), usersFile.getPath()));
+						GroupManager.logger.log(Level.WARNING, String.format(Messages.getString("WorldDatHolder.WARN_NO_GROUP_STATED"), thisUserNode.get("group").toString(), thisUser.getLastName(), dataHolder.getDefaultGroup().getName(), usersFile.getPath()));
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/storage/Yaml.java:
```diff
@@ -844,7 +844,7 @@ public void saveGroups(WorldDataHolder dataHolder) {
 
 			if (dataHolder.getDefaultGroup() == null) {
-				GroupManager.logger.severe(Messages.getString("WorldDatHolder.WARN_NO_DEFAULT_GROUP") + dataHolder.getName());
+				GroupManager.logger.log(Level.SEVERE, Messages.getString("WorldDatHolder.WARN_NO_DEFAULT_GROUP") + dataHolder.getName());
```
Functionally the same.

ElgarL commented on src/org/anjocaido/groupmanager/utils/BukkitWrapper.java:
```diff
@@ -157,7 +157,7 @@ public OfflinePlayer getOfflinePlayer(UUID uid) {
 	 * This list is not sorted in any particular order. If an exact match is found, the returned list will only contain a single result.
 	 * 
-	 * @param the (partial) name to match
+	 * @param name to match
```
This is a partial name match.

l1ttleO replied:
IntelliJ refused to stop complaining otherwise, but I'll add the `(partial)` after the `name`

ElgarL commented on src/org/anjocaido/groupmanager/utils/Tasks.java:
```diff
@@ -183,54 +182,54 @@ public static String getStringListInString(List<String> list) {
-		String out = arr[0];
+		StringBuilder out = Optional.ofNullable(arr[0]).map(StringBuilder::new).orElse(null);
 		for (int i = 1; i < arr.length; i++)
-			out += separator + arr[i];
-		return out;
+			out = (out == null ? new StringBuilder("null") : out).append(separator).append(arr[i]);
+		return out == null ? null : out.toString();
```
This is overly complex for a simple method. I really dislike Optional.
StringBuilder yes, Optional is excessive. I prefer simple readable code.

l1ttleO replied:
I guess it used Optional because I used null-safe when converting to a StringBuilder, will fix

l1ttleO commented:
I replaced the loggers for consistency, because it differed every line or two. If you care about the import, I will replace it with fully classified name

I removed the synchronization because IntelliJ complained about syncing on a non-final field

l1ttleO commented on src/org/anjocaido/groupmanager/dataholder/worlds/MirrorsMap.java:
```diff
@@ -80,15 +80,7 @@ public void parseMirrors() {
-				 */
-				
-			} else {
+			} else if (child != null) {
```
will do

l1ttleO commented on src/org/anjocaido/groupmanager/dataholder/worlds/ChildMirrors.java:
```diff
@@ -19,8 +19,8 @@
 	 */
-	private Map<String, String> groupsMirror = new TreeMap<>();
-	private Map<String, String> usersMirror = new TreeMap<>();
+	private final Map<String, String> groupsMirror = new TreeMap<>();
+	private final Map<String, String> usersMirror = new TreeMap<>();
```
I'll check tomorrow, but in general, reloads are near guaranteed to cause problems and server admins are advised to avoid reloading as much as possible

l1ttleO commented on src/org/anjocaido/groupmanager/dataholder/WorldDataHolder.java:
```diff
@@ -525,23 +516,17 @@ public boolean purgeTimedPermissions() {
 			}
 		}
 
-		synchronized (users) {
-
-			for (User user : getUsers().values()) {
-				if (user.removeExpired()) {
-					setUsersChanged(true);
-					expired = true;
-				}
```
I don't have an environment where I can do prolonged and extensive testing, so either this would need to be reverted or it would be published as a potentially highly unstable build

l1ttleO commented on src/org/anjocaido/groupmanager/dataholder/WorldDataHolder.java:
```diff
@@ -525,23 +516,17 @@ public boolean purgeTimedPermissions() {
 
 		boolean expired = false;
 
-		synchronized (groups) {
-
-			for (Group group : getGroups().values()) {
-				if (group.removeExpired()) {
-					setGroupsChanged(true);
-					expired = true;
-				}
```
I don't have an environment where I can do prolonged and extensive testing, so either this would need to be reverted or it would be published as a potentially highly unstable build

l1ttleO commented on src/org/anjocaido/groupmanager/dataholder/WorldDataHolder.java:
```diff
@@ -510,10 +503,8 @@ public void removeUsersChangedFlag() {
 	public void removeGroupsChangedFlag() {
 
 		setGroupsChanged(false);
-		synchronized (groups) {
-			for (Group g : getGroups().values()) {
-				g.flagAsSaved();
-			}
```
I don't have an environment where I can do prolonged and extensive testing, so either this would need to be reverted or it would be published as a potentially highly unstable build

l1ttleO commented on src/org/anjocaido/groupmanager/dataholder/WorldDataHolder.java:
```diff
@@ -497,10 +492,8 @@ public boolean haveGroupsChanged() {
 	public void removeUsersChangedFlag() {
 
 		setUsersChanged(false);
-		synchronized (users) {
-			for (User u : getUsers().values()) {
-				u.flagAsSaved();
-			}
```
I don't have an environment where I can do prolonged and extensive testing, so either this would need to be reverted or it would be published as a potentially highly unstable build

l1ttleO commented on src/org/anjocaido/groupmanager/dataholder/WorldDataHolder.java:
```diff
@@ -481,11 +478,9 @@ public boolean haveGroupsChanged() {
 		if (groups.isGroupsChanged()) {
 			return true;
 		}
-		synchronized (groups) {
-			for (Group g : groups.getGroups().values()) {
-				if (g.isChanged()) {
-					return true;
-				}
```
I don't have an environment where I can do prolonged and extensive testing, so either this would need to be reverted or it would be published as a potentially highly unstable build

l1ttleO commented on src/org/anjocaido/groupmanager/dataholder/WorldDataHolder.java:
```diff
@@ -454,13 +453,11 @@ public boolean haveUsersChanged() {
 		if (users.isUsersChanged()) {
 			return true;
 		}
-		synchronized (users) {
-			for (User u : users.getUsers().values()) {
-				if (u.isChanged()) {
-					return true;
```
I don't have an environment where I can do prolonged and extensive testing, so either this would need to be reverted or it would be published as a potentially highly unstable build

l1ttleO commented on src/org/anjocaido/groupmanager/data/User.java:
```diff
@@ -401,13 +397,13 @@ public User updatePlayer() {
 	public Player getBukkitPlayer() {
 
-		return BukkitWrapper.getInstance().getPlayer(getLastName());
+		return BukkitWrapper.getInstance(..getPlayer(getLastName());
```
yeah, I fixed it in a later commit

l1ttleO commented on src/org/anjocaido/groupmanager/data/User.java:
```diff
@@ -326,12 +327,7 @@ public boolean removeSubGroup(Group subGroup) {
-			} else {
-				groupList.add(g);
-			}
+			groupList.add(g);
```
will do

l1ttleO commented on src/org/anjocaido/groupmanager/data/DataUnit.java:
```diff
@@ -201,10 +202,7 @@ public void flagAsSaved() {
-		synchronized (permissions) {
-			return permissions.containsKey(permission);
-		}
+		return permissions.containsKey(permission);
```
I removed the sync because it was assumed that syncing wouldn't be safe either because we're syncing on a non-final field

l1ttleO commented on src/org/anjocaido/groupmanager/data/DataUnit.java:
```diff
@@ -43,7 +44,7 @@
 	private long timeStamp = 0;
 
-	private Map<String, Long> permissions = Collections.synchronizedSortedMap(new TreeMap<>(new StringPermissionComparator()));
+	private final Map<String, Long> permissions = Collections.synchronizedSortedMap(new TreeMap<>(new StringPermissionComparator()));
```
I don't have an environment where I can do prolonged and extensive testing, so either this would need to be reverted or it would be published as a potentially highly unstable build

ElgarL reviewed (changes requested):
Closer.

ElgarL commented on src/org/anjocaido/groupmanager/dataholder/worlds/MirrorsMap.java:
```diff
@@ -12,7 +12,7 @@
 
 /**
- * @author ElgarL
+ * @author Level
```
lol

l1ttleO replied:
wait what

ElgarL commented on src/org/anjocaido/groupmanager/GroupManager.java:
```diff
@@ -185,18 +186,21 @@ public void onEnable(boolean restarting) {
 			selectedWorlds = new HashMap<>();
 			lastError = ""; //$NON-NLS-1$
 
+			if (!restarting) {
```
This section is still not correct. Things are out of order. Like loading GlobalGroups. This whole section change (other than the logger changes) should probably be scrapped.
